### PR TITLE
Create ledger queries sample from marbles02 sample

### PR DESCRIPTION
**Pull request by stephyee**
Signed-off-by: Tiffany Harris <tiffany.harris@ibm.com>

### CODE REVIEW

File: asset-transfer-ledger-queries/chaincode-go/asset_transfer_ledger_chaincode.go [hunk 1]
```diff
@@ -0,0 +1,589 @@
...
+	Timestamp           time.Time `json:"timestamp"`
+	FetchedRecordsCount int       `json:"fetchedRecordsCount"`
+	Bookmark            string    `json:"bookmark"`
+}
+
+// ============================================================
+// CreateAsset - create a new asset, store into chaincode state
+// ============================================================
+func (t *SimpleChaincode) CreateAsset(ctx contractapi.TransactionContextInterface, ID, color, owner string,
+	size, appraisedValue int) error {
```
Comment (DereckLuo) [inline asset-transfer-ledger-queries/chaincode-go/asset_transfer_ledger_chaincode.go:119]: nit: new line seems a bit strange here. If the parameters are too long, might want to put each parameter in its own line.

File: asset-transfer-ledger-queries/chaincode-go/asset_transfer_ledger_chaincode.go [hunk 2]
```diff
@@ -0,0 +1,589 @@
...
+
+	// ==== Asset saved and indexed. Return success ====
+	return nil
+}
+
+// ===============================================
+// ReadAsset - read a asset from chaincode state
+// ===============================================
+func (t *SimpleChaincode) ReadAsset(ctx contractapi.TransactionContextInterface, name string) (*asset, error) {
+	valAsbytes, err := ctx.GetStub().GetState(name) //get the asset from chaincode state
```
Comment (DereckLuo) [inline asset-transfer-ledger-queries/chaincode-go/asset_transfer_ledger_chaincode.go:169]: `valAsbytes` might not be a good name here. Can use `assetJSON` or maybe `serializedAsset`. 
(applies to all the other `valAsbytes`)

File: asset-transfer-ledger-queries/chaincode-go/asset_transfer_ledger_chaincode.go [hunk 3]
```diff
@@ -0,0 +1,589 @@
...
+// DeleteAsset - remove a asset key/value pair from state
+// ==================================================
+func (t *SimpleChaincode) DeleteAsset(ctx contractapi.TransactionContextInterface, assetName string) error {
+	var jsonResp string
+	var assetJSON asset
+
+	// to maintain the color~name index, we need to read the asset first and get its color
+	valAsbytes, err := ctx.GetStub().GetState(assetName) //get the asset from chaincode state
+	if err != nil {
+		jsonResp = "{\"Error\":\"Failed to get state for " + assetName + "\"}"
```
Comment (DereckLuo) [inline asset-transfer-ledger-queries/chaincode-go/asset_transfer_ledger_chaincode.go:195]: not sure what's the purpose to use jsonResp for error formatting. Can probably just use `fmt.Errorf`.

File: asset-transfer-ledger-queries/chaincode-go/go.mod [hunk 4]
```diff
@@ -0,0 +1,13 @@
+module github.com/hyperledger/fabric-samples/chaincode/marbles02/go
+
+go 1.12
```
Comment (DereckLuo) [inline asset-transfer-ledger-queries/chaincode-go/go.mod:3]: can bump this to 1.14

File: asset-transfer-ledger-queries/chaincode-go/go.mod [hunk 5]
```diff
@@ -0,0 +1,13 @@
+module github.com/hyperledger/fabric-samples/chaincode/marbles02/go
```
Comment (DereckLuo) [inline asset-transfer-ledger-queries/chaincode-go/go.mod:1]: should this be `github.com/hyperledger/fabric-samples/asset-transfer-ledger-queries/chaincode-go`

File: asset-transfer-ledger-queries/chaincode-go/go.mod (hunk 4 above)
Comment (stephyee) [reply asset-transfer-ledger-queries/chaincode-go/go.mod:3]: 🛠️ Ah I copied this over from marbles02.

File: asset-transfer-ledger-queries/chaincode-go/go.mod (hunk 5 above)
Comment (stephyee) [reply asset-transfer-ledger-queries/chaincode-go/go.mod:1]: 🛠️ Ah I copied this over from marbles02.

File: asset-transfer-ledger-queries/chaincode-go/asset_transfer_ledger_chaincode.go (hunk 3 above)
Comment (stephyee) [reply asset-transfer-ledger-queries/chaincode-go/asset_transfer_ledger_chaincode.go:195]: 🛠️  this is from when a pb.Response was being returned.

File: asset-transfer-ledger-queries/chaincode-go/asset_transfer_ledger_chaincode.go (hunk 2 above)
Comment (stephyee) [reply asset-transfer-ledger-queries/chaincode-go/asset_transfer_ledger_chaincode.go:169]: 🛠️

File: asset-transfer-ledger-queries/chaincode-go/asset_transfer_ledger_chaincode.go [hunk 6]
```diff
@@ -0,0 +1,584 @@
...
+	Record              *asset
+	TxId                string    `json:"txId"`
+	Timestamp           time.Time `json:"timestamp"`
+	FetchedRecordsCount int       `json:"fetchedRecordsCount"`
+	Bookmark            string    `json:"bookmark"`
+}
+
+// ============================================================
+// CreateAsset - create a new asset, store into chaincode state
+// ============================================================
```
Comment (denyeart) [inline asset-transfer-ledger-queries/chaincode-go/asset_transfer_ledger_chaincode.go:117]: You can remove the equals signs (throughout) and use a normal Go comment, e.g.
// CreateAsset creates a new asset and stores into chaincode state.

File: asset-transfer-ledger-queries/chaincode-go/asset_transfer_ledger_chaincode.go [hunk 7]
```diff
@@ -0,0 +1,584 @@
...
+	ctx.GetStub().PutState(colorNameIndexKey, value)
+
+	// ==== Asset saved and indexed. Return success ====
+	return nil
+}
+
+// ===============================================
+// ReadAsset - read a asset from chaincode state
+// ===============================================
+func (t *SimpleChaincode) ReadAsset(ctx contractapi.TransactionContextInterface, name string) (*asset, error) {
```
Comment (denyeart) [inline asset-transfer-ledger-queries/chaincode-go/asset_transfer_ledger_chaincode.go:167]: Change asset 'name' to 'ID' throughout.

File: asset-transfer-ledger-queries/chaincode-go/asset_transfer_ledger_chaincode.go [hunk 8]
```diff
@@ -0,0 +1,584 @@
...
+	assetJSON, err := ctx.GetStub().GetState(name) //get the asset from chaincode state
```
Comment (denyeart) [inline asset-transfer-ledger-queries/chaincode-go/asset_transfer_ledger_chaincode.go:168]: You can remove many of these obvious comments, comments are only needed for non-obvious things.

File: asset-transfer-ledger-queries/chaincode-go/asset_transfer_ledger_chaincode.go [hunk 9]
```diff
@@ -0,0 +1,584 @@
...
+	if err != nil {
+		return fmt.Errorf("Failed to get asset: " + err.Error())
+	} else if exists {
+		return fmt.Errorf("This asset already exists: " + ID)
+	}
+
+	// ==== Create asset object and marshal to JSON ====
+	objectType := "asset"
+	asset := &asset{objectType, ID, color, size, owner, appraisedValue}
+	assetJSONasBytes, err := json.Marshal(asset)
```
Comment (denyeart) [inline asset-transfer-ledger-queries/chaincode-go/asset_transfer_ledger_chaincode.go:131]: assetJSON would be a sufficient variable name.

File: asset-transfer-ledger-queries/chaincode-go/asset_transfer_ledger_chaincode.go [hunk 10]
```diff
@@ -0,0 +1,584 @@
...
+// =========================================================================================
+func (t *SimpleChaincode) QueryAssetsByOwner(ctx contractapi.TransactionContextInterface, owner string) ([]*QueryResult, error) {
+	queryString := fmt.Sprintf("{\"selector\":{\"docType\":\"asset\",\"owner\":\"%s\"}}", owner)
+
+	queryResults, err := getQueryResultForQueryString(ctx, queryString)
+	if err != nil {
+		return nil, err
+	}
+
+	return queryResults, nil
```
Comment (denyeart) [inline asset-transfer-ledger-queries/chaincode-go/asset_transfer_ledger_chaincode.go:382]: can just return getQueryResultForQueryString(ctx, queryString)

File: asset-transfer-ledger-queries/chaincode-go/asset_transfer_ledger_chaincode.go [hunk 11]
```diff
@@ -0,0 +1,584 @@
...
+// If this is not desired, follow the QueryAssetsForOwner example for parameterized queries.
+// Only available on state databases that support rich query (e.g. CouchDB)
+// =========================================================================================
+func (t *SimpleChaincode) QueryAssets(ctx contractapi.TransactionContextInterface, queryString string) ([]*QueryResult, error) {
+	queryResults, err := getQueryResultForQueryString(ctx, queryString)
+	if err != nil {
+		return nil, err
+	}
+
+	return queryResults, nil
```
Comment (denyeart) [inline asset-transfer-ledger-queries/chaincode-go/asset_transfer_ledger_chaincode.go:398]: can just return getQueryResultForQueryString(ctx, queryString)

File: asset-transfer-ledger-queries/chaincode-go/asset_transfer_ledger_chaincode.go [hunk 12]
```diff
@@ -0,0 +1,584 @@
...
+		return nil, err
+	}
+	defer resultsIterator.Close()
+
+	buffer, err := constructQueryResponseFromIterator(resultsIterator)
+	if err != nil {
+		return nil, err
+	}
+
+	return buffer, nil
```
Comment (denyeart) [inline asset-transfer-ledger-queries/chaincode-go/asset_transfer_ledger_chaincode.go:418]: can just return constructQueryResponseFromIterator(resultsIterator)

File: asset-transfer-ledger-queries/chaincode-go/asset_transfer_ledger_chaincode.go [hunk 13]
```diff
@@ -0,0 +1,584 @@
...
```
Comment (denyeart) [inline asset-transfer-ledger-queries/chaincode-go/asset_transfer_ledger_chaincode.go:454]: can just return constructQueryResponseFromIterator(resultsIterator)

File: asset-transfer-ledger-queries/chaincode-go/asset_transfer_ledger_chaincode.go [hunk 14]
```diff
@@ -0,0 +1,584 @@
...
+// Paginated queries are only valid for read only transactions.
+// =========================================================================================
+func (t *SimpleChaincode) QueryAssetsWithPagination(ctx contractapi.TransactionContextInterface, queryString,
+	bookmark string, pageSize int) ([]*QueryResult, error) {
+	queryResults, err := getQueryResultForQueryStringWithPagination(ctx, queryString, int32(pageSize), bookmark)
+	if err != nil {
+		return nil, err
+	}
+
+	return queryResults, nil
```
Comment (denyeart) [inline asset-transfer-ledger-queries/chaincode-go/asset_transfer_ledger_chaincode.go:473]: Just return again

File: asset-transfer-ledger-queries/chaincode-go/asset_transfer_ledger_chaincode.go [hunk 15]
```diff
@@ -0,0 +1,584 @@
...
+		return nil, err
+	}
+	defer resultsIterator.Close()
+
+	buffer, err := constructQueryResponseFromIterator(resultsIterator)
+	if err != nil {
+		return nil, err
+	}
+
+	return buffer, nil
```
Comment (denyeart) [inline asset-transfer-ledger-queries/chaincode-go/asset_transfer_ledger_chaincode.go:493]: Just return again

File: asset-transfer-ledger-queries/chaincode-go/asset_transfer_ledger_chaincode.go [hunk 16]
```diff
@@ -0,0 +1,584 @@
...
+}
+
+func (t *SimpleChaincode) GetAssetHistory(ctx contractapi.TransactionContextInterface, args []string) ([]QueryResult, error) {
+
+	if len(args) < 1 {
+		return nil, fmt.Errorf("Incorrect number of arguments. Expecting 1")
+	}
+
+	assetName := args[0]
```
Comment (denyeart) [inline asset-transfer-ledger-queries/chaincode-go/asset_transfer_ledger_chaincode.go:502]: Pass asset ID instead of args

**Comment by stephyee** [comment]
@denyeart Are you in favor of adding a readme? I know there are a few references to marbles02 in the docs. However most of the query examples exist in the code comments.

**Comment by denyeart** [comment]
> @denyeart Are you in favor of adding a readme? I know there are a few references to marbles02 in the docs. However most of the query examples exist in the code comments.

The plan is to update the existing CouchDB tutorial for a step by step index and json query. So we could use a very small readme that mentions what content is included, but I don't think we need step-by-step details in the readme.

**Pull request merged**